File: backend/app/services/llm.py
"""OCR processing service using LLM (e.g., Google Gemini or Gemma)."""

import json
import logging
from datetime import datetime
from typing import Optional, Dict, Any
from pathlib import Path

from PIL import Image
from google import genai
from google.genai import types

from ..config import get_settings
from .categorizer import VALID_CATEGORIES

settings = get_settings()
logger = logging.getLogger(__name__)

# Initialize LLM Client
client = None
if settings.google_api_key:
    client = genai.Client(api_key=settings.google_api_key)
else:
    logger.warning("GOOGLE_API_KEY not found in settings. Gemini features will be disabled.")

def extract_json_from_response(text: str) -> Optional[Dict[str, Any]]:
    """Clean markdown code blocks and parse JSON."""
    try:
        # Remove markdown code formatting if present
        text = text.strip()
        
        # Try to find the first '{' and last '}' to isolate JSON
        start_idx = text.find('{')
        end_idx = text.rfind('}')
        
        if start_idx != -1 and end_idx != -1 and end_idx > start_idx:
            text = text[start_idx:end_idx + 1]
        
        return json.loads(text)
    except json.JSONDecodeError:
        logger.error(f"Failed to parse JSON from LLM response: {text}")
        return None

async def process_receipt_image(image_path: str) -> dict:
    """
    Process a receipt image using Google Gemini API.
    """
    if not client:
        return {
            "vendor": None,
            "amount": None,
            "date": None,
            "currency": "USD",
            "raw_text": "Error: Gemini client not initialized (missing API key)",
            "confidence": 0.0,
        }

    if not Path(image_path).exists():
        raise FileNotFoundError(f"Image not found: {image_path}")

    try:
        # Load image for Gemini
        img = Image.open(image_path)
        
        categories_str = ", ".join(VALID_CATEGORIES)
        
        prompt = f"""
Act as an advanced OCR and data extraction assistant. Analyze the provided receipt image and extract specific data points into a structured JSON format.

### Extraction Instructions:
1. **Vendor**: Identify the official name of the store or service provider.
2. **Date**: Extract the transaction date. Normalize to "YYYY-MM-DD".
3. **Amount**: Locate the final "Total" or "Amount Due". Exclude sub-totals.
4. **Currency**: Extract the 3-letter ISO 4217 currency code (e.g., USD, EUR, GBP, INR). Convert symbols if necessary (e.g., "$" -> "USD", "€" -> "EUR", "₹" -> "INR").
5. **Category**: Assign the most relevant category from this list: [{categories_str}].

### Output Schema (Strict JSON):
{{
    "vendor": "string or null",
    "date": "string or null",
    "amount": number or null,
    "currency": "string or null",
    "category": "string or null"
}}

Respond ONLY with valid JSON matching this schema. Do not include any other text.
"""        
        
        logger.info(f"Calling LLM API with model {settings.llm_model}")
        
        # Call Gemini using google-genai SDK
        # We use a synchronous call in an executor or just run it since it's likely blocking if not using async client
        # google-genai client.models.generate_content is synchronous by default.
        # However, we are in an async function.
        
        response = client.models.generate_content(
            model=settings.llm_model,
            contents=[prompt, img],
            config=types.GenerateContentConfig(
                temperature=0.1,
                max_output_tokens=1024, # Increased for potentially longer reasoning
            )
        )
        
        content = response.text
        logger.info(f"LLM Response: {content}")
        
        parsed_data = extract_json_from_response(content) or {}
        
        # Convert date string to object
        date_obj = None
        if parsed_data.get("date"):
            try:
                date_obj = datetime.strptime(parsed_data["date"], "%Y-%m-%d").date()
            except ValueError:
                pass

        return {
            "vendor": parsed_data.get("vendor"),
            "amount": parsed_data.get("amount"),
            "date": date_obj,
            "currency": parsed_data.get("currency", "USD"),
            "category": parsed_data.get("category"),
            "raw_text": content, # Store full LLM response as raw text
            "confidence": 1.0 
        }

    except Exception as e:
        logger.error(f"Gemini processing failed: {e}")
        return {
            "confidence": 0.0,
        }

async def process_receipt_audio(audio_path: str) -> dict:
    """
    Process a receipt audio recording using Google GenAI (brand: Gemini/Gemma).
    """
    if not client:
        return {
            "vendor": None,
            "amount": None,
            "date": None,
            "currency": "USD",
            "raw_text": "Error: GenAI client not initialized (missing API key)",
            "confidence": 0.0,
        }

    if not Path(audio_path).exists():
        raise FileNotFoundError(f"Audio file not found: {audio_path}")

    try:
        # Load audio file
        with open(audio_path, "rb") as f:
            audio_bytes = f.read()

        # Determine mime type based on extension
        ext = Path(audio_path).suffix.lower()
        mime_type = "audio/wav" if ext == ".wav" else "audio/webm"
        
        categories_str = ", ".join(VALID_CATEGORIES)
        
        prompt = f"""
Act as an advanced receipt data extraction assistant. Listen to the audio recording where a user describes a purchase and extract specific data points into a structured JSON format.

### Extraction Instructions:
1. **Vendor**: Identify the store or service provider mentioned.
2. **Date**: Extract the transaction date if mentioned (e.g., "yesterday", "last friday", "on January 12th"). Convert relative dates to absolute YYYY-MM-DD based on today's date: {datetime.now().strftime("%Y-%m-%d")}.
3. **Amount**: Extract the total amount spent.
4. **Currency**: Extract the currency if mentioned, otherwise default to USD.
5. **Category**: Assign the most relevant category from this list: [{categories_str}].

### Output Schema (Strict JSON):
{{
    "vendor": "string or null",
    "date": "string or null",
    "amount": number or null,
    "currency": "string or null",
    "category": "string or null"
}}

Respond ONLY with valid JSON matching this schema. Do not include any other text.
"""        
        
        # Use Gemini Flash Latest as stable fallback
        audio_model = "models/gemini-flash-latest"
        logger.info(f"Calling LLM API with model {audio_model} for audio processing")
        
        response = client.models.generate_content(
            model=audio_model,
            contents=[
                prompt,
                types.Part.from_bytes(data=audio_bytes, mime_type=mime_type)
            ],
            config=types.GenerateContentConfig(
                temperature=0.1,
                max_output_tokens=1024,
            )
        )
        
        content = response.text
        logger.info(f"LLM Audio Response: {content}")
        
        parsed_data = extract_json_from_response(content) or {}
        
        # Convert date string to object
        date_obj = None
        if parsed_data.get("date"):
            try:
                date_obj = datetime.strptime(parsed_data["date"], "%Y-%m-%d").date()
            except ValueError:
                pass

        return {
            "vendor": parsed_data.get("vendor"),
            "amount": parsed_data.get("amount"),
            "date": date_obj,
            "currency": parsed_data.get("currency", "USD"),
            "category": parsed_data.get("category"),
            "raw_text": content,
            "confidence": 1.0 
        }

    except Exception as e:
        logger.error(f"GenAI audio processing failed: {e}")
        return {
            "vendor": None,
            "amount": None,
            "date": None,
            "currency": "USD",
            "raw_text": f"Error: {str(e)}",
            "confidence": 0.0,
        }

File: backend-go/internal/services/llm.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	"google.golang.org/genai"
)

type ExtractedReceipt struct {
	Vendor     *string  `json:"vendor"`
	Date       *string  `json:"date"`
	Amount     *float64 `json:"amount"`
	Currency   *string  `json:"currency"`
	Category   *string  `json:"category"`
	RawText    string
	Confidence float64
}

var validCategoriesStr = strings.Join(ValidCategories, ", ")

var imagePrompt = fmt.Sprintf(`Act as an advanced OCR and data extraction assistant. Analyze the provided receipt image and extract specific data points into a structured JSON format.

### Extraction Instructions:
1. **Vendor**: Identify the official name of the store or service provider.
2. **Date**: Extract the transaction date. Normalize to "YYYY-MM-DD".
3. **Amount**: Locate the final "Total" or "Amount Due". Exclude sub-totals.
4. **Currency**: Extract the 3-letter ISO 4217 currency code (e.g., USD, EUR, GBP, INR).
5. **Category**: Assign the most relevant category from this list: [%s].

### Output Schema (Strict JSON):
{"vendor":"string or null","date":"string or null","amount":null,"currency":"string or null","category":"string or null"}

Respond ONLY with valid JSON matching this schema.`, validCategoriesStr)

func newGenAIClient(ctx context.Context, apiKey string) (*genai.Client, error) {
	return genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  apiKey,
		Backend: genai.BackendGeminiAPI,
	})
}

func extractJSON(text string) map[string]interface{} {
	text = strings.TrimSpace(text)
	start := strings.Index(text, "{")
	end := strings.LastIndex(text, "}")
	if start == -1 || end == -1 || end <= start {
		return nil
	}
	text = text[start : end+1]
	var result map[string]interface{}
	if err := json.Unmarshal([]byte(text), &result); err != nil {
		log.Printf("Failed to parse JSON from LLM response: %v", err)
		return nil
	}
	return result
}

func parseDate(s string) *time.Time {
	if s == "" {
		return nil
	}
	t, err := time.Parse("2006-01-02", s)
	if err != nil {
		return nil
	}
	return &t
}

func strPtr(s string) *string  { return &s }
func f64Ptr(f float64) *float64 { return &f }

func ProcessReceiptImage(imagePath string, apiKey string, model string) (*ExtractedReceipt, error) {
	if apiKey == "" {
		return &ExtractedReceipt{RawText: "Error: missing API key", Confidence: 0}, nil
	}
	imgBytes, err := os.ReadFile(imagePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read image: %w", err)
	}

	// Detect mime type from extension
	mime := "image/jpeg"
	low := strings.ToLower(imagePath)
	if strings.HasSuffix(low, ".png") {
		mime = "image/png"
	} else if strings.HasSuffix(low, ".webp") {
		mime = "image/webp"
	}

	ctx := context.Background()
	client, err := newGenAIClient(ctx, apiKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create genai client: %w", err)
	}

	parts := []*genai.Part{
		genai.NewPartFromText(imagePrompt),
		genai.NewPartFromBytes(imgBytes, mime),
	}

	log.Printf("Calling LLM API with model %s", model)
	resp, err := client.Models.GenerateContent(ctx, model, []*genai.Content{genai.NewContentFromParts(parts, "user")}, &genai.GenerateContentConfig{
		Temperature: genai.Ptr[float32](0.1),
		MaxOutputTokens: 1024,
	})
	if err != nil {
		return nil, fmt.Errorf("LLM call failed: %w", err)
	}

	text := resp.Text()
	log.Printf("LLM Response: %s", text)
	data := extractJSON(text)
	result := &ExtractedReceipt{RawText: text, Confidence: 1.0}
	if data != nil {
		if v, ok := data["vendor"].(string); ok && v != "" {
			result.Vendor = strPtr(v)
		}
		if v, ok := data["amount"].(float64); ok {
			result.Amount = f64Ptr(v)
		}
		if v, ok := data["currency"].(string); ok && v != "" {
			result.Currency = strPtr(v)
		}
		if v, ok := data["category"].(string); ok && v != "" {
			result.Category = strPtr(v)
		}
		if v, ok := data["date"].(string); ok && v != "" {
			result.Date = strPtr(v)
		}
	}
	return result, nil
}

func ProcessReceiptAudio(audioPath string, apiKey string) (*ExtractedReceipt, error) {
	if apiKey == "" {
		return &ExtractedReceipt{RawText: "Error: missing API key", Confidence: 0}, nil
	}
	audioBytes, err := os.ReadFile(audioPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read audio: %w", err)
	}

	mime := "audio/webm"
	low := strings.ToLower(audioPath)
	if strings.HasSuffix(low, ".wav") {
		mime = "audio/wav"
	} else if strings.HasSuffix(low, ".mp3") {
		mime = "audio/mpeg"
	} else if strings.HasSuffix(low, ".m4a") {
		mime = "audio/mp4"
	}

	audioModel := "models/gemini-flash-latest"
	prompt := fmt.Sprintf(`Act as an advanced receipt data extraction assistant. Listen to the audio recording where a user describes a purchase and extract specific data points into a structured JSON format.

### Extraction Instructions:
1. Vendor: identify the store or service mentioned.
2. Date: extract the transaction date if mentioned. Convert relative dates using today: %s. Normalize to YYYY-MM-DD.
3. Amount: extract the total amount spent.
4. Currency: extract the currency if mentioned, otherwise default to USD.
5. Category: assign the most relevant category from: [%s].

Output ONLY valid JSON: {"vendor":null,"date":null,"amount":null,"currency":null,"category":null}`,
		time.Now().Format("2006-01-02"), validCategoriesStr)

	ctx := context.Background()
	client, err := newGenAIClient(ctx, apiKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create genai client: %w", err)
	}

	parts := []*genai.Part{
		{Text: prompt},
		{InlineData: &genai.Blob{Data: audioBytes, MIMEType: mime}},
	}

	log.Printf("Calling LLM API with model %s for audio", audioModel)
	resp, err := client.Models.GenerateContent(ctx, audioModel, []*genai.Content{genai.NewContentFromParts(parts, "user")}, &genai.GenerateContentConfig{
		Temperature: genai.Ptr[float32](0.1),
		MaxOutputTokens: 1024,
	})
	if err != nil {
		return nil, fmt.Errorf("audio LLM call failed: %w", err)
	}

	text := resp.Text()
	log.Printf("LLM Audio Response: %s", text)
	data := extractJSON(text)
	result := &ExtractedReceipt{RawText: text, Confidence: 1.0}
	if data != nil {
		if v, ok := data["vendor"].(string); ok && v != "" {
			result.Vendor = strPtr(v)
		}
		if v, ok := data["amount"].(float64); ok {
			result.Amount = f64Ptr(v)
		}
		if v, ok := data["currency"].(string); ok && v != "" {
			result.Currency = strPtr(v)
		}
		if v, ok := data["category"].(string); ok && v != "" {
			result.Category = strPtr(v)
		}
		if v, ok := data["date"].(string); ok && v != "" {
			result.Date = strPtr(v)
		}
	}
	return result, nil
}
